File: rcnb.js
/*!
  RCNB.js 
  Copyright (c) 2019-2022 Coxxs
  License: https://github.com/rcnbapp/RCNB.js/blob/master/LICENSE
*/

// char
const cr = 'rRŔŕŖŗŘřƦȐȑȒȓɌɍ'
const cc = 'cCĆćĈĉĊċČčƇƈÇȻȼ'
const cn = 'nNŃńŅņŇňƝƞÑǸǹȠȵ'
const cb = 'bBƀƁƃƄƅßÞþ'

// size
const sr = cr.length,
  sc = cc.length,
  sn = cn.length,
  sb = cb.length
const src = sr * sc
const snb = sn * sb
const scnb = sc * snb

function _div(a, b) {
  return Math.floor(a / b)
}

function _encodeByte(i) {
  if (i > 0xFF) throw new Error('rc/nb overflow')
  if (i > 0x7F) {
    i = i & 0x7F
    return cn.charAt(_div(i, sb)) + cb.charAt(i % sb)
  }
  return cr.charAt(_div(i, sc)) + cc.charAt(i % sc)
}

function _encodeShort(i) {
  if (i > 0xFFFF) throw new Error('rcnb overflow')
  let reverse = false
  if (i > 0x7FFF) {
    reverse = true
    i = i & 0x7FFF
  }
  let char = [
    _div(i, scnb),
    _div(i % scnb, snb),
    _div(i % snb, sb),
    i % sb
  ]
  char = [cr[char[0]], cc[char[1]], cn[char[2]], cb[char[3]]]
  if (reverse) {
    return char[2] + char[3] + char[0] + char[1]
  }
  return char.join('')
}

function _decodeByte(c) {
  let nb = false
  let idx = [cr.indexOf(c.charAt(0)), cc.indexOf(c.charAt(1))]
  if (idx[0] < 0 || idx[1] < 0) {
    idx = [cn.indexOf(c.charAt(0)), cb.indexOf(c.charAt(1))]
    nb = true
  }
  if (idx[0] < 0 || idx[1] < 0) throw new Error('not rc/nb')
  const result = nb ? idx[0] * sb + idx[1] : idx[0] * sc + idx[1]
  if (result > 0x7F) throw new Error('rc/nb overflow')
  return nb ? result | 0x80 : result
}

function _decodeShort(c) {
  let idx
  const reverse = cr.indexOf(c.charAt(0)) < 0
  if (!reverse) {
    idx = [cr.indexOf(c.charAt(0)), cc.indexOf(c.charAt(1)), cn.indexOf(c.charAt(2)), cb.indexOf(c.charAt(3))]
  } else {
    idx = [cr.indexOf(c.charAt(2)), cc.indexOf(c.charAt(3)), cn.indexOf(c.charAt(0)), cb.indexOf(c.charAt(1))]
  }
  if (idx[0] < 0 || idx[1] < 0 || idx[2] < 0 || idx[3] < 0) throw new Error('not rcnb')
  const result = idx[0] * scnb + idx[1] * snb + idx[2] * sb + idx[3]
  if (result > 0x7FFF) throw new Error('rcnb overflow')
  return reverse ? result | 0x8000 : result
}

export function encode(arr) {
  if (!(arr instanceof Uint8Array)) {
    throw new TypeError('argument is not an instance of Uint8Array')
  }
  let str = ''
  // encode every 2 bytes
  for (let i = 0; i < (arr.length >> 1); i++) {
    str += _encodeShort((arr[i * 2] << 8) | arr[i * 2 + 1])
  }
  // encode tailing byte
  if (arr.length & 1) str += _encodeByte(arr[arr.length - 1])
  return str
}

export function decode(str) {
  if (str.length & 1) throw new Error('invalid length')
  const arr = []
  // decode every 2 bytes (1 rcnb = 2 bytes)
  for (let i = 0; i < (str.length >> 2); i++) {
    const short = _decodeShort(str.substr(i * 4, 4))
    arr.push(short >> 8)
    arr.push(short & 0xFF)
  }
  // decode tailing byte (1 rc / 1 nb = 1 byte)
  if (str.length & 2) arr.push(_decodeByte(str.substr(-2, 2)))
  return Uint8Array.from(arr)
}

export default { encode, decode }

File: rcnb.py
#!/usr/bin/env python3
#-*- encoding=utf-8 -*-

import math
import sys

cr = 'rRŔŕŖŗŘřƦȐȑȒȓɌɍ'
cc = 'cCĆćĈĉĊċČčƇƈÇȻȼ'
cn = 'nNŃńŅņŇňƝƞÑǸǹȠȵ'
cb = 'bBƀƁƃƄƅßÞþ'
# print(cr)
# print(cc)
# print(cn)
# print(cb)

sr = len(cr)
sc = len(cc)
sn = len(cn)
sb = len(cb)
# print(sr)
# print(sc)
# print(sn)
# print(sb)

src = sr * sc
snb = sn * sb
scnb = sc * snb
# print(src)
# print(snb)
# print(scnb)

def _div(a, b):
	return math.floor(a / b)
# print(_div(3, 4))
# print(_div(9, 4))

def _encodeByte(i):
	if i > 0xFF:
		print('rc/nb overflow')
		sys.exit(-1)
	if i > 0x7F:
		i = i & 0x7F
		return cn[_div(i, sb)] + cb[i % sb]
	return cr[_div(i, sc)] + cc[i % sc]
# t1 = _encodeByte(0x80)
# t2 = _encodeByte(0x20)
# print(t1)
# print(t2)

def _encodeShort(i):
	if i > 0xFFFF:
		print('rcnb overflow')
		sys.exit(-1)
	reverse = False
	if i > 0x7FFF:
		reverse = True
		i = i & 0x7FFF
	ch = [_div(i, scnb), _div(i % scnb, snb), _div(i % snb, sb), i % sb]
# 	print(ch)
	res = [cr[ch[0]], cc[ch[1]], cn[ch[2]], cb[ch[3]]]
# 	print(res)
	if reverse:
		return res[2] + res[3] + res[0] + res[1]
	return ''.join(res)
# t3 = _encodeShort(0x8000)
# t4 = _encodeShort(0x2000)
# print(t3)
# print(t4)

def _decodeByte(c):
	nb = False
	idx = [cr.find(c[0]), cc.find(c[1])]
# 	print(idx)
	if idx[0] < 0 or idx[1] < 0:
		idx = [cn.find(c[0]), cb.find(c[1])]
		nb = True
	if idx[0] < 0 or idx[1] < 0:
		print('not rc/nb')
		sys.exit(-1)
	result = (idx[0] * sb + idx[1]) if nb else (idx[0] * sc + idx[1])
	if result > 0x7F:
		print('rc/nb overflow')
		sys.exit(-1)
# 	print(nb)
	return (result | 0x80) if nb else result
# print(_decodeByte(t1))
# print(_decodeByte(t2))

def _decodeShort(c):
	idx = 0
	reverse = cr.find(c[0]) < 0
	if not reverse:
		idx = [cr.find(c[0]), cc.find(c[1]), cn.find(c[2]), cb.find(c[3])]
	else:
		idx = [cr.find(c[2]), cc.find(c[3]), cn.find(c[0]), cb.find(c[1])]
	if idx[0] < 0 or idx[1] < 0 or idx[2] < 0 or idx[3] < 0:
		print('not rcnb')
		sys.exit(-1)
	result = idx[0] * scnb + idx[1] * snb + idx[2] * sb + idx[3]
	if result > 0x7FFF:
		print('rcnb overflow')
		sys.exit(-1)
# 	print(reverse)
	return (reverse | 0x8000) if reverse else result
# print(_decodeShort(t3))
# print(_decodeShort(t4))

def encode(m):
	res = ''
	for i in range(len(m) >> 1):
		res += _encodeShort(ord(m[2 * i]) << 8 | ord(m[2 * i + 1]))
	if len(m) & 1 == 1:
		res += _encodeByte(ord(m[-1]))
	return res

def decode(c):
	if len(c) & 1 == 1:
		print('invalid length')
		sys.exit(-1)
	res = ''
	for i in range(len(c) >> 2):
		short = _decodeShort(c[4 * i:4 * i + 4])
		res += chr(short >> 8)
		res += chr(short & 0xFF)
	if len(c) & 2 == 2:
		res += chr(_decodeByte(c[-2:]))
	return res

# s = 'hello'
# es = encode(s)
# print(es)
# ds = decode(es)
# print(ds)

if __name__ == '__main__':
	if len(sys.argv) != 3:
		print('Usage:', sys.argv[0], 'encode [PLAIN]')
		print('      ', sys.argv[0], 'decode [ENCRYPTED]')
		sys.exit(-1)
	if sys.argv[1] == 'encode':
		res = encode(sys.argv[2])
		print('Result:', res)
		sys.exit(0)
	if sys.argv[1] == 'decode':
		res = decode(sys.argv[2])
		print('Result:', res)
		sys.exit(0)
	else:
		print('arg wrong')
		sys.exit(-1)

